compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,511 @@
+    //   2) In the backward sweep, only hoist dmas for which
+    //   (source.memory > target.memory), i.e. moving data away from AIE cores.
+    //
+    // These last checks in theory limit the hoisting detection capability, but
+    // should be valid.
+    forOp.walk<WalkOrder::PostOrder, Iterator>([&](AMDAIE::DmaCpyNdOp dmaOp) {
+      if (llvm::any_of(dmaOp->getOpOperands(), [&](OpOperand &operand) {
+            return dependsOnLoop(operand);
+          })) {
+        addDependencies(dmaOp);
+        return WalkResult::advance();
+      }
+
+      uint64_t sourceMemspace =
+          dmaOp.getSourceObjectFifo().getMemorySpaceAsUInt();
+      uint64_t targetMemspace =
+          dmaOp.getTargetObjectFifo().getMemorySpaceAsUInt();
+      if (std::is_same<Iterator, ForwardIterator>::value &&
+          !dependencies.contains(dmaOp.getSourceObjectFifo()) &&
+          sourceMemspace < targetMemspace) {
+        rewriter.moveOpBefore(dmaOp, forOp);
+        hoistHappened = true;
+        return WalkResult::advance();
+      } else if (std::is_same<Iterator, ReverseIterator>::value &&
+                 !dependencies.contains(dmaOp.getTargetObjectFifo()) &&
+                 sourceMemspace > targetMemspace) {
+        rewriter.moveOpAfter(dmaOp, forOp);
+        hoistHappened = true;
+        return WalkResult::advance();
+      }
+
+      // If this dma op can't be hoisted due to dependencies, keep adding new
+      // dependencies.
+      addDependencies(dmaOp);
+      return WalkResult::advance();
+    });
+    if (hoistHappened) return success();
+    return failure();
+  }
+
+  /// Unroll the for loop, skipping the first iteration. While unrolling, create
+  /// new clones of `LogicalObjectFifoFromMemrefOp` so they can be distributed
+  /// onto multiple physical locations later.
+  LogicalResult loopUnrollAndDistributeLogicalObjectFifos(
+      RewriterBase &rewriter, scf::ForOp forOp) const {
+    Block *loopBodyBlock = forOp.getBody();
+    OpBuilder builder = OpBuilder::atBlockTerminator(loopBodyBlock);
+
+    // Keep a pointer to the last non-terminator operation in the original block
+    // so that we know what to clone (since we are doing this in-place).
+    Block::iterator srcBlockEnd = std::prev(loopBodyBlock->end(), 2);
+
+    // Update loop bounds
+    int64_t lbInt = getConstantIntValue(forOp.getLowerBound()).value();
+    int64_t ubInt = getConstantIntValue(forOp.getUpperBound()).value();
+    int64_t stepInt = getConstantIntValue(forOp.getStep()).value();
+    if (lbInt != 0 || stepInt != 1) return failure();
+    if (stepInt == (ubInt - lbInt)) return failure();
+    Value forOpIV = forOp.getInductionVar();
+    forOp.setLowerBound(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 0));
+    forOp.setUpperBound(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 1));
+    forOp.setStep(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 1));
```

Comment:
Guess I'm failing to understand part of `forOp.set<>Bound(...)`.

If we have made sure that `lb = 0` and `step = 1` then why do we need to set those bounds again? Shouldn't we just do that for the `ub` ?

Reply:
Yes, this is indeed not needed for now because of those checks.